### PR TITLE
Color map transparency

### PR DESCRIPTION
**Pull request by Haydelj**
Implemented colormap transparency using user-defined points addressing issue #1361

### CODE REVIEW

File: src/Core/Datatypes/ColorMap.cc [hunk 1]
```diff
@@ -153,7 +154,7 @@ double ColorMap::getTransformedValue(double f) const
   static bool x = true;
```
Comment (dcwhite) [inline src/Core/Datatypes/ColorMap.cc]: This weird nonsense code was put in a long time ago due to a strange crash which I can't even remember. Let's try removing the whole block

File: src/Core/Datatypes/ColorMap.cc [hunk 2]
```diff
@@ -194,48 +195,81 @@ ColorRGB ColorMap::getColorMapVal(double v) const
   //now grab the RGB
   auto colorWithoutAlpha = color_->getColorMapVal(f);
   //TODO:
-  //return applyAlpha(f, colorWithoutAlpha);
-  return colorWithoutAlpha;
+  return applyAlpha(f, colorWithoutAlpha);
+  //return colorWithoutAlpha;
```
Comment (dcwhite) [inline src/Core/Datatypes/ColorMap.cc]:
```suggestion
  
```

File: src/Core/Datatypes/ColorMap.cc [hunk 3]
```diff
@@ -194,48 +195,81 @@ ColorRGB ColorMap::getColorMapVal(double v) const
...
-...easy
 
+ColorRGB ColorMap::applyAlpha(double transformed, ColorRGB colorWithoutAlpha) const
+{
+  double a = alpha(transformed);
+  return ColorRGB(colorWithoutAlpha.r(), colorWithoutAlpha.g(), colorWithoutAlpha.b(), a);
+}
 
-double alpha(double transformedValue)
+double ColorMap::alpha(double transformedValue) const
```
Comment (dcwhite) [inline src/Core/Datatypes/ColorMap.cc]: This function could be easily unit tested

File: src/Core/Datatypes/ColorMap.h [hunk 4]
```diff
@@ -51,7 +51,8 @@ namespace Datatypes {
     explicit ColorMap(ColorMapStrategyHandle color,
       const std::string& name = "Rainbow", const size_t resolution = 256,
                         const double shift = 0.0, const bool invert = false,
-                        const double rescale_scale = .5, const double rescale_shift = 1.);
+                        const double rescale_scale = .5, const double rescale_shift = 1.0,
+                        const std::vector<double>& alphaPoints = std::vector<double>());
```
Comment (dcwhite) [inline src/Core/Datatypes/ColorMap.h]:
```suggestion
                        const std::vector<double>& alphaPoints = {});
```

File: src/Core/Datatypes/ColorMap.h [hunk 5]
```diff
@@ -112,7 +116,8 @@ namespace Datatypes {
    // See explanation for defaults above in ColorMap Constructor
     static ColorMapHandle create(const std::string& name = "Rainbow", const size_t &resolution = 256,
                                     const double &shift = 0.0, const bool &invert = false,
-                                    const double &rescale_scale = .5, const double &rescale_shift = 1.);
+                                    const double &rescale_scale = .5, const double &rescale_shift = 1.0,
+                                    const std::vector<double>& alphaPoints = std::vector<double>());
```
Comment (dcwhite) [inline src/Core/Datatypes/ColorMap.h]:
```suggestion
                                    const std::vector<double>& alphaPoints = {});
```

File: src/Core/Datatypes/ColorMap.cc (hunk 3 above)
Comment (dcwhite) [reply src/Core/Datatypes/ColorMap.cc]: ...which would earn lots more XP...

File: src/Interface/Modules/Visualization/CreateStandardColorMapDialog.cc [hunk 6]
```diff
@@ -264,9 +249,8 @@ void AlphaFunctionManager::pushToState()
       Variable::List alphaPointsVec;
       //strip endpoints before saving user-added points
```
Comment (dcwhite) [inline src/Interface/Modules/Visualization/CreateStandardColorMapDialog.cc]: Delete this comment now right? This is one reason why comments can be bad...

File: src/Interface/Modules/Visualization/CreateStandardColorMapDialog.cc [hunk 7]
```diff
@@ -317,12 +302,11 @@ void AlphaFunctionManager::updateAlphaFunction()
 
   for (int i = 0; i < static_cast<int>(alphaFunction_.size()); ++i)
   {
-    if (i > 0 && i < alphaFunction_.size() - 1)
+    if (false && i > 0 && i < alphaFunction_.size() - 1)
```
Comment (dcwhite) [inline src/Interface/Modules/Visualization/CreateStandardColorMapDialog.cc]: ???

**Comment by dcwhite** [comment]
Waiting on #1987

File: src/Interface/Modules/Visualization/CreateStandardColorMapDialog.cc (hunk 7 above)
Comment (Haydelj) [reply src/Interface/Modules/Visualization/CreateStandardColorMapDialog.cc]: haha whoops was testing to see if this code even mattered. Seems like it doesn't will take a closer look before removing entierly.

File: src/Core/Datatypes/ColorMap.cc (hunk 1 above)
Comment (dcwhite) [reply src/Core/Datatypes/ColorMap.cc]: https://github.com/SCIInstitute/SCIRun/commit/cf0fb5aa4a5c438e96e3a42545cfec8e913ae5ed#diff-e6d3c5861036ceaf0b1ed068fc84a547

Comment (Haydelj) [reply src/Core/Datatypes/ColorMap.cc]: I removed the code and it built and ran just fine on mac and windows

File: src/Core/Application/Version.cc [hunk 8]
```diff
@@ -0,0 +1,32 @@
+/*
+   For more information, please see: http://software.sci.utah.edu
+
+   The MIT License
+
+   Copyright (c) 2015 Scientific Computing and Imaging Institute,
+   University of Utah.
+
```
Comment (dcwhite) [inline src/Core/Application/Version.cc]: This file should be deleted--it lives in the bin folder now

**Comment by dcwhite** [comment]
@Haydelj Merge conflict alert

**Comment by dcwhite** [comment]
Merge today

**Comment by jessdtate** [comment]
The alpha selecting widget in the module UI moves as expected, but doesn't seem to actually change the transparency.

**Comment by Haydelj** [comment]
> The alpha selecting widget in the module UI moves as expected, but doesn't seem to actually change the transparency.

It turned out that the rescale color map module was the culprit it's fixed now.

**Comment by dcwhite** [comment]
I'm seeing some triangles that appear and then disappear while rotating with mixed transparency. Here's an example
t2
t3
[j1.srn5.zip](https://github.com/SCIInstitute/SCIRun/files/3703723/j1.srn5.zip)

**Comment by dcwhite** [comment]
There were also many of these: `Unable to find uniform with name: uTransparency in shader with ID: 10`

**Comment by jessdtate** [comment]
I think the triangles are a result of the sorting, which needs some work still.

**Pull request merged**